motif/matchComp.go

```diff
@@ -0,0 +1,288 @@
+		refPos = fasta.AlnPosToRefPosCounter(records[0], alnPos, lastRefPos, lastAlnPos)
+		//first we need a key for the alnPos. We can either make a new key or update our current key
+		if needNewKey {
+			currRefKey, alnPos, couldGetNewKey = getNewKey(records[0], alnPos, len(pm.Mat[0]))
+			refPos = fasta.AlnPosToRefPosCounter(records[0], alnPos, lastRefPos, lastAlnPos)
+			lastRefPos, lastAlnPos = refPos, alnPos
+			if !couldGetNewKey {
+				break //this means we've run out of windows on the reference sequence.
+			}
+			needNewKey = false
+		} else {
+			switch records[0].Seq[alnPos] {
+			case dna.N:
+				needNewKey = true
+				continue
+			case dna.Gap:
+				continue
+			case dna.A:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 0
+				currRefKey = currRefKey & bitMask
+			case dna.C:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 1
+				currRefKey = currRefKey & bitMask
+			case dna.G:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 2
+				currRefKey = currRefKey & bitMask
+			case dna.T:
+				currRefKey = currRefKey << 2
+				currRefKey = currRefKey | 3
+				currRefKey = currRefKey & bitMask
+			default:
+				log.Fatalf("Unrecognized base.")
+			}
+		}
+
+		// now we check if the current key is a significant motif hit
+		if currRefScore, inKmerHash = kmerHash[currRefKey]; inKmerHash {
+			//fmt.Printf("CurrRefKey: %v. CurrRefScore: %v.\n", currRefKey, currRefScore)
+			minResidual = math.Inf(1)
+			minResidualAltScore = 0
+			for currAltStart = numbers.Max(alnPos-len(pm.Mat[0])-residualWindowSize+1, 0); currAltStart <= alnPos+residualWindowSize-len(pm.Mat[0])+1; currAltStart++ {
+				currAltScore, currAltEnd, couldScoreSequence = ScoreWindow(pm, records[1].Seq, currAltStart)
+				//fmt.Printf("AlnPos: %v. currAltStart: %v. CurrAltEnd: %v. CurrAltScore: %v.\n", alnPos, currAltStart, currAltEnd, currAltScore)
+				if !couldScoreSequence {
+					break
+				}
+				currResidual = math.Abs(currRefScore - currAltScore)
+				if currResidual < minResidual {
+					minResidual = currResidual
+					minResidualAltScore = currAltScore
+				}
+				if !enforceStrandMatch {
+					currAltScore, currAltEnd, couldScoreSequence = ScoreWindow(revCompPm, records[1].Seq, currAltStart)
+					//fmt.Printf("AlnPos: %v. currAltStart: %v. CurrAltEnd: %v. CurrAltScore: %v.\n", alnPos, currAltStart, currAltEnd, currAltScore)
+					if !couldScoreSequence {
+						break
+					}
+					currResidual = math.Abs(currRefScore - currAltScore)
+					if currResidual < minResidual {
+						minResidual = currResidual
+						minResidualAltScore = currAltScore
+					}
+				}
+				altEndsConsidered[currAltEnd] = true
+			}
+			if outputAsProportion {
+				currRefScore = currRefScore / consensusScore
+				minResidualAltScore = minResidualAltScore / consensusScore
+				minResidual = math.Abs(currRefScore - minResidualAltScore)
+			}
+			if minResidual >= residualFilter {
+				currBed = bed.Bed{
```

Comment:
In motif/match.go you have the below code, which I thought was right (chr1 0 10, motif is positions 0-9 aka length is 10, currPos is 10 after scanning the genome window, so chrEnd is 10, chrEnd is 10-10=0). Is that consistent with the code here? Here, is refPos 9 after scanning the genome window (maybe doesn't get ++ at the end of a loop?), and chrEnd is 9+1=10, chrStart is 9-10+1=0)?
```
					currBed = bed.Bed{Chrom: records[currChrom].Name,
						ChromStart:        currPos - motifLen,
						ChromEnd:          currPos,
						Name:              motifName,
						Score:             0,
						Strand:            strand,
						FieldsInitialized: 7,
						Annotation:        []string{fmt.Sprintf("%f", currScore)}}
```